presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -868,6 +868,20 @@ public void testTermOrder()
                                 values("A"))));
     }
 
+    @Test
+    public void testCastToTimestampWithTimeZone()
+    {
+        assertPlan("SELECT * FROM (VALUES TIMESTAMP '2017-01-02 00:00:00') t(a) WHERE a > TIMESTAMP '1970-01-01 00:00:00 UTC'",
+                anyTree(
+                        filter("A >= \"@$literal$@52QIVV94KJAG6012H0D1N0119RA6LCSSVNKM8B6JTILU3JAK21IEDGHNG60HPHG7PTGMGG1EK97JIUKSEI3OIA16H860JI4Q7NTACGH10Q5H8GCI8V7V3LAS2BG4DEIRONQPD9KEBO20HDMK79EBJB19D1TONRF80EHHKRHIOF9N2ADTIE7ERBEI4NEPQ100TE8GM1LJ3MA8JC9LUM7SBOBN\"(BIGINT '0')",
```

Comment:
`$literal$` -> https://github.com/prestosql/presto/issues/5683

I am concerned about `>=` here. This looks like a bug in Unwrap, some assumption that is not true for timestamp with time zone. @martint am i right?

Reply:
it's session zone dependent (so for example i get `>` if i add:
```
 Session.builder(getQueryRunner().getDefaultSession())
                        .setTimeZoneKey(TimeZoneKey.getTimeZoneKey("Europe/Warsaw"))
                        .build(),
```
)

the default zone is Apia. I'd expect to easily find values that pass for Apia with `>`...

Reply:
Yeah, that's a bug. The problem is that the implicit coercion between `timestamp` -> `timestamp with time zone` is ill-defined due to its dependency on the session timezone. In particular:

`CAST(TIMESTAMP '1970-01-01 00:00:00 UTC' AS TIMESTAMP) = TIMESTAMP '1970-01-01 00:00:00'`

 but 

`CAST(TIMESTAMP '1970-01-01 00:00:00' AS TIMESTAMP WITH TIME ZONE) <> TIMESTAMP '1970-01-01 00:00:00 UTC'`

(i.e, the conversion doesn't roundtrip)

We can add an exception to the Unwrap optimization, but we should also consider removing the implicit coercion between those types.

Reply:
https://github.com/prestosql/presto/issues/5685